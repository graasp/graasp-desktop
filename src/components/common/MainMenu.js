import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import SaveIcon from '@material-ui/icons/Save';
import CodeIcon from '@material-ui/icons/Code';
import ListItemText from '@material-ui/core/ListItemText';
import List from '@material-ui/core/List';
import SearchIcon from '@material-ui/icons/Search';
import Language from '@material-ui/icons/Language';
import PublishIcon from '@material-ui/icons/Publish';
import SettingsIcon from '@material-ui/icons/Settings';
import { Online } from 'react-detect-offline';
import { withTranslation } from 'react-i18next';
import {
  HOME_PATH,
  LOAD_SPACE_PATH,
  SETTINGS_PATH,
  SPACES_NEARBY_PATH,
  VISIT_PATH,
  DEVELOPER_PATH,
} from '../../config/paths';

export class MainMenu extends Component {
  static propTypes = {
    t: PropTypes.func.isRequired,
    developerMode: PropTypes.bool.isRequired,
    history: PropTypes.shape({ replace: PropTypes.func.isRequired }).isRequired,
    match: PropTypes.shape({ path: PropTypes.string.isRequired }).isRequired,
  };

  handleClick = path => {
    const {
      history: { replace },
    } = this.props;
    if (path) {
      replace(path);
    } else {
      // default to home
      replace(HOME_PATH);
    }
  };

  renderDeveloperMode() {
    const {
      developerMode,
      t,
      match: { path },
    } = this.props;

    if (developerMode) {
      return (
        <MenuItem
          onClick={() => this.handleClick(DEVELOPER_PATH)}
          selected={path === DEVELOPER_PATH}
          button
        >
          <ListItemIcon>
            <CodeIcon />
          </ListItemIcon>
          <ListItemText primary={t('Developer')} />
        </MenuItem>
      );
    }
    return null;
  }

  render() {
    const {
      match: { path },
      t,
    } = this.props;
    return (
      <List>
        <MenuItem
          onClick={() => this.handleClick(HOME_PATH)}
          button
          selected={path === HOME_PATH}
        >
          <ListItemIcon>
            <SaveIcon />
          </ListItemIcon>
          <ListItemText primary={t('Saved Spaces')} />
        </MenuItem>
        <Online>
          <MenuItem
            onClick={() => this.handleClick(SPACES_NEARBY_PATH)}
            button
            selected={path === SPACES_NEARBY_PATH}
          >
            <ListItemIcon>
              <SearchIcon />
            </ListItemIcon>
            <ListItemText primary={t('Spaces Nearby')} />
          </MenuItem>
          <MenuItem
            onClick={() => this.handleClick(VISIT_PATH)}
            button
            selected={path === VISIT_PATH}
          >
            <ListItemIcon>
              <Language />
            </ListItemIcon>
            <ListItemText primary={t('Visit a Space')} />
          </MenuItem>
        </Online>
        <MenuItem
          onClick={() => this.handleClick(LOAD_SPACE_PATH)}
          button
          selected={path === LOAD_SPACE_PATH}
        >
          <ListItemIcon>
            <PublishIcon />
          </ListItemIcon>
          <ListItemText primary={t('Load')} />
        </MenuItem>
        <MenuItem
          onClick={() => this.handleClick(SETTINGS_PATH)}
          button
          selected={path === SETTINGS_PATH}
        >
          <ListItemIcon>
            <SettingsIcon />
          </ListItemIcon>
          <ListItemText primary={t('Settings')} />
        </MenuItem>
        {this.renderDeveloperMode()}
      </List>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  developerMode: User.getIn(['current', 'developerMode']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const ConnectedComponent = connect(mapStateToProps)(MainMenu);

const TranslatedComponent = withTranslation()(ConnectedComponent);

export default withRouter(TranslatedComponent);
