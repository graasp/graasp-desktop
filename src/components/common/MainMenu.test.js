import React from 'react';
import { shallow } from 'enzyme';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import List from '@material-ui/core/List';
import { Online } from 'react-detect-offline';
import { MainMenu } from './MainMenu';
import {
  HOME_PATH,
  SPACES_NEARBY_PATH,
  VISIT_PATH,
  LOAD_SPACE_PATH,
  SETTINGS_PATH,
  DEVELOPER_PATH,
} from '../../config/paths';

const createMainMenuProps = (developerMode, path) => {
  return {
    t: text => text,
    developerMode,
    history: { replace: jest.fn() },
    match: { path },
  };
};
describe('<MainMenu />', () => {
  describe('<MainMenu /> with developerMode = false', () => {
    let wrapper;

    beforeAll(() => {
      const props = createMainMenuProps(false, HOME_PATH);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<MainMenu {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <List /> component', () => {
      expect(wrapper.find(List)).toHaveLength(1);
    });

    it('renders one <Online /> component containing 2 <MenuItems />', () => {
      expect(wrapper.find(Online)).toHaveLength(1);
      expect(wrapper.find(Online).find(MenuItem)).toHaveLength(2);
    });

    it('renders five <MenuItem /> components', () => {
      expect(wrapper.find(MenuItem)).toHaveLength(5);
    });

    it('renders five <MenuItem /> components with unique icons', () => {
      const menuItems = wrapper.find(MenuItem);
      expect(menuItems).toHaveLength(5);

      const icons = menuItems.find(ListItemIcon);
      expect(icons).toHaveLength(5);

      const uniqueIcons = icons
        .map(icon => icon.dive())
        .filter((item, i, ar) => ar.indexOf(item) === i);
      expect(icons.length).toBe(uniqueIcons.length);
    });
  });

  describe('<MainMenu /> with developerMode = true', () => {
    let wrapper;

    beforeAll(() => {
      const props = createMainMenuProps(true, HOME_PATH);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<MainMenu {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <List /> component', () => {
      expect(wrapper.find(List)).toHaveLength(1);
    });

    it('renders one <Online /> component containing 2 <MenuItems />', () => {
      expect(wrapper.find(Online)).toHaveLength(1);
      expect(wrapper.find(Online).find(MenuItem)).toHaveLength(2);
    });

    it('renders six <MenuItem /> components', () => {
      expect(wrapper.find(MenuItem)).toHaveLength(6);
    });

    it('renders six <MenuItem /> components with unique icons', () => {
      const menuItems = wrapper.find(MenuItem);
      expect(menuItems).toHaveLength(6);

      const icons = menuItems.find(ListItemIcon);
      expect(icons).toHaveLength(6);

      const uniqueIcons = icons
        .map(icon => icon.dive())
        .filter((item, i, ar) => ar.indexOf(item) === i);
      expect(icons.length).toBe(uniqueIcons.length);
    });
  });

  describe.each([
    [true, HOME_PATH, 'Saved Spaces'],
    [true, SPACES_NEARBY_PATH, 'Spaces Nearby'],
    [true, VISIT_PATH, 'Visit a Space'],
    [true, LOAD_SPACE_PATH, 'Load'],
    [true, SETTINGS_PATH, 'Settings'],
    [true, DEVELOPER_PATH, 'Developer'],
  ])('<MainMenu /> selects one MenuItem', (developerMode, path, text) => {
    it(`select path=${path} (developerMode = ${developerMode})`, () => {
      const props = createMainMenuProps(developerMode, path);
      // eslint-disable-next-line react/jsx-props-no-spreading
      const wrapper = shallow(<MainMenu {...props} />);
      const selectedMenuItem = wrapper.find({ selected: true });
      expect(selectedMenuItem).toHaveLength(1);
      expect(selectedMenuItem.find(ListItemText).prop('primary')).toEqual(text);
    });
  });
});
