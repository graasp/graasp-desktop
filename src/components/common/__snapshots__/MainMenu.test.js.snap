// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainMenu /> <MainMenu /> with developerMode = false renders correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={true}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SaveIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Saved Spaces"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <SearchIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Spaces Nearby"
      />
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <LanguageIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Visit a Space"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <PublishIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Load"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SettingsIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Settings"
    />
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(List))>
`;

exports[`<MainMenu /> <MainMenu /> with developerMode = true renders correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={true}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SaveIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Saved Spaces"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <SearchIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Spaces Nearby"
      />
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <LanguageIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Visit a Space"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <PublishIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Load"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SettingsIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Settings"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <CodeIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Developer"
    />
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(List))>
`;
